src/SipHash.chpl

```diff
@@ -55,6 +56,17 @@ module SipHash {
             (p[7]: uint(64) << 56));
   }
 
+  private inline proc XTO64_LE(in x: ?t) {
+    var y: uint(64);
+    if isSubtype(t, c_ptr) {
```

Comment:
I agree with your intuition here. I actually started with a `select`/`when` block here and in another location below, but for some reason the Chapel compiler was complaining about it, and switching to an equivalent `if`/`else` block made it compile. @ronawho are there any known issues with using `select` for compile-time info like this?

Reply:
It's not clear to me how this would work with the `select` statement. From https://chapel-lang.org/docs/language/spec/statements.html#the-select-statement:

> The expression that follows the keyword select, the select expression, is evaluated once and its value is then compared with the list of case expressions following each when keyword. These values are compared using the equality operator `==`. If the expressions cannot be compared with the equality operator, a compile-time error is generated.

But I don't see an obvious way to do `==` against `t` here.

Reply:
@ronawho ah, sorry, I had confused the snippet above with this code here
https://github.com/Bears-R-Us/arkouda/blob/32d18ef7ed165903722a591540e5237c93d9362b/src/SipHash.chpl#L197-L212

For that code, I had originally tried something like
```chpl
select (stride, isSubtype(msg.type, c_ptr)) {
  when (8, true) { m = U8TO64_LE(msg + pos); }
  when (8, false) { m = U8TO64_LE(msg, pos..#stride); }
  when (1, true) { m = XTO64_LE(msg + pos); }
  when (1, false) { m = XTO64_LE(msg[pos]); }
  otherwise { compilerError(); }
}
```

But the compiler was complaining about it. I forget the error message, but I'm working from home tomorrow, so I can try to reproduce it then.

Reply:
Ah, I _think_ that should work, but I'm not positive. If you can reproduce the error let me know.

Reply:
@ronawho Here's a reproducer for this error. The code below compiles (and runs) with `-smode=Mode.ifelse`, but when I try to compile with `-smode=Mode.selectwhen`, it throws the compilerError in the `otherwise` block. The error only happens when the `select` statement operates on both `stride` and `isSubtype`. I tried a `select` with each of them separately and it compiled fine.

```chpl
private use CPtr;
enum Mode {
  selectwhen,
  ifelse
}
config param mode = Mode.selectwhen;

proc dothing(x, param eltBytes) where (mode == Mode.selectwhen) {
  param stride: int = if (eltBytes == 1) then 8 else 1;
  select (stride, isSubtype(x.type, c_ptr)) {
  when (8, true) { writeln("c_ptr(uint8)"); }
  when (8, false) { writeln("[] uint(8)"); }
  when (1, true) { writeln("c_ptr(int64)"); }
  when (1, false) { writeln("[] int(64)"); }
  otherwise { compilerError(stride:string + " " + x.type:string); }
  }
}

proc dothing(x, param eltBytes) where (mode == Mode.ifelse) {
  param stride: int = if (eltBytes == 1) then 8 else 1;
  if (stride == 8) {
    if isSubtype(x.type, c_ptr) {
      writeln("c_ptr(uint8)");
    } else {
      writeln("[] uint(8)");
    }
  } else if (stride == 1) {
    if isSubtype(x.type, c_ptr) {
      writeln("c_ptr(int64)");
    } else {
      writeln("[] int(64)");
    }
  } else {
    compilerError("unsupported type");
  }
}

proc main() {
  var msg64: [0..#10] int;
  var msg8: [0..#10] int;
  dothing(msg64, 8);
  dothing(c_ptrTo(msg64), 8);
  dothing(msg8, 1);
  dothing(c_ptrTo(msg8), 1);
}
```

Reply:
Ah, so I think what happens here is that we don't have `param` tuples so once you put your params into a tuple the compiler can no longer reason about them so it turns into an execution time check. If you rewrite your `ifelse` version to use tuples it hits the compilerError too:

```chpl
proc dothing(x, param eltBytes) where (mode == Mode.ifelse) {
  param stride: int = if (eltBytes == 1) then 8 else 1;
  param isSub = isSubtype(x.type, c_ptr);
  if (stride, isSub) == (8, true) {
    writeln("c_ptr(uint8)");
  } else if (stride, isSub) == (8, false) {
      writeln("[] uint(8)");
  } else if (stride, isSub) == (1, true) {
    writeln("c_ptr(int64)");
  } else if (stride, isSub) == (1, false) {
    writeln("[] int(64)");
  } else {
    compilerError("unsupported type");
  }
}
```

Reply:
Oh, that makes sense. I think I'll just leave this as-is then.